### ISSUE TITLE
Clang 14's `<stdatomic.h>` is incompatible with MSVC's `<stdatomic.h>` :radioactive:

### ISSUE THREAD

**Issue by StephanTLavavej** [microsoft/STL#2862]
This repros with Clang 14, when the STL's tests are compiled with the official toolset (instead of the microsoft/STL repo - the difference is in the header locations, because Clang overrides toolset headers).

`tests/std/tests/P0943R6_stdatomic_h/test.compile.pass.cpp` and the many tests including `stl/inc/__msvc_all_public_headers.hpp` fail due to Clang's `<stdatomic.h>`, like this:

```
RUN.PM line 750:  Executing: "clang-cl -m32  /nologo /Od /W4 /w14061 /w14242 /w14265 /w14582 /w14583 /w14587 /w14588 /w14749 /w14841 /w14842 /w15038 /w15214 /w15215 /w15216 /w15217 /sdl /WX /Zc:strictStrings /D_ENABLE_STL_INTERNAL_CHECK /bigobj /FIforce_include.hpp /w14365 /D_ENFORCE_FACET_SPECIALIZATIONS=1 /D_STL_CALL_ABORT_INSTEAD_OF_INVALID_PARAMETER -fno-ms-compatibility -fno-delayed-template-parsing /EHsc /MD /std:c++14 /w14640 /Zc:threadSafeInit-   .\test.cpp  /c".
In file included from .\test.cpp:4:
In file included from D:\Contest\ca1z2udg.zif\binaries\x86chk\inc\__msvc_all_public_headers.hpp:146:
D:\Contest\ca1z2udg.zif\src\tools\LLVM\LLVM14\x86\lib\clang\14.0.5\include\stdatomic.h(70,6): error: conflicting types for 'atomic_thread_fence'
void atomic_thread_fence(memory_order);
     ^
D:\Contest\ca1z2udg.zif\binaries\x86chk\inc\atomic(166,24): note: previous definition is here
extern "C" inline void atomic_thread_fence(const memory_order _Order) noexcept {
                       ^
[...]
```

As a workaround, we're going to avoid including `<stdatomic.h>` for Clang, but solving the root cause would be better.

**Comment by fsb4000** [comment]
They have many macros: 
```c++
#define atomic_thread_fence(order) __c11_atomic_thread_fence(order)
#define atomic_fetch_xor_explicit __c11_atomic_fetch_xor
....
```
and we have real functions instead of macros in `<atomic>`:
```c++
extern "C" inline void atomic_thread_fence(const memory_order _Order) noexcept {
    ::_Atomic_thread_fence(static_cast<unsigned int>(_Order));
}
....
```
and then macro replaces to 
```
extern "C" inline void __c11_atomic_thread_fence(const memory_order _Order) noexcept {
    ::_Atomic_thread_fence(static_cast<unsigned int>(_Order));
}
```
but `__c11_atomic_thread_fence` is builtin.
So I think Clang just should include `<stdatomic.h>` if it is C++ and MSVC and use their `<stdatomic>` only for C:
https://github.com/llvm/llvm-project/blob/6aff1b7b3ca71e048375a85a11e611527da6a45c/clang/lib/Headers/stdatomic.h#L13-L22

I will prepare a patch.

**Comment by fsb4000** [comment]
They landed the patch : https://github.com/llvm/llvm-project/commit/ba49d39b20cc5358da28af2ac82bd336028780bc
Thanks @mordante and @AaronBallman ❤️

**Comment by h-vetinari** [comment]
Speaking of `<stdatomic.h>` - the [conformance page](https://docs.microsoft.com/en-us/cpp/overview/visual-cpp-language-conformance#c-standard-library-features-1) says visual studio does not support it, and I cannot [find](https://developercommunity.visualstudio.com/search?q=stdatomic.h) a corresponding issue in the dev community for it - is there a plan to bring this to conformance eventually?

**Comment by fsb4000** [comment]
Yes, DevCom-1204057

**Comment by h-vetinari** [comment]
> Yes, [DevCom-1204057](https://developercommunity.visualstudio.com/t/1204057)

Thanks! Guess I was searching wrong (under "visual studio" instead of under "C++").

**Comment by mordante** [comment]
> They landed the patch : [llvm/llvm-project@ba49d39](https://github.com/llvm/llvm-project/commit/ba49d39b20cc5358da28af2ac82bd336028780bc) Thanks @mordante and @AaronBallman heart

You're welcome. FYI the patch landed before LLVM 15 was branched. So the fix will be in Clang-15.

**Comment by fsb4000** [comment]
`__cplusplus` and `_STL_LANG` are `202002L` for `clang-cl` and `/std:c++latest`.
This is why the test is not enabled: https://github.com/microsoft/STL/blob/c873cf01323f2f37ce3021f344d905dc84411fe3/tests/std/tests/P0943R6_stdatomic_h/test.compile.pass.cpp#L4
The test works if I define the macro manually
![изображение](https://user-images.githubusercontent.com/4289847/198893058-1117bc02-1aa3-4e00-80ba-e4ef99e9feed.png)
